src/test/javascript/spec/component/exam/manage/programming-exam-diff.component.spec.ts

```diff
@@ -0,0 +1,137 @@
+            of({
+                id: 2,
+            } as ProgrammingExerciseGitDiffReport),
+        );
+        component.previousSubmission = { id: 1 };
+        component.currentSubmission = { id: 2 };
+        component.loadGitDiffReport();
+        expect(diffForSubmissionsSpy).toHaveBeenCalledExactlyOnceWith(3, 1, 2);
+        expect(diffForSubmissionWithTemplateSpy).not.toHaveBeenCalled();
+        expect(component.addedLineCount).toBe(3);
+        expect(component.removedLineCount).toBe(1);
+    });
+
+    it('should call getDiffReportForSubmissionWithTemplate when loading diff report if previous submission is undefined', () => {
+        const diffForSubmissionsSpy = jest.spyOn(programmingExerciseService, 'getDiffReportForSubmissions').mockReturnValue(of({ id: 1 } as ProgrammingExerciseGitDiffReport));
+        const diffForSubmissionWithTemplateSpy = jest
+            .spyOn(programmingExerciseService, 'getDiffReportForSubmissionWithTemplate')
+            .mockReturnValue(of({ id: 2 } as ProgrammingExerciseGitDiffReport));
+        component.previousSubmission = undefined;
+        component.currentSubmission = { id: 2 };
+        component.exercise = { id: 3 } as ProgrammingExercise;
+        component.loadGitDiffReport();
+        expect(diffForSubmissionWithTemplateSpy).toHaveBeenCalledOnce();
+        expect(diffForSubmissionsSpy).not.toHaveBeenCalled();
+    });
+
+    it('should open the modal when showGitDiff is called', () => {
+        const modalServiceSpy = jest.spyOn(modal, 'open');
+        component.exercise = { id: 1 } as ProgrammingExercise;
+        component.showGitDiff();
+        expect(modalServiceSpy).toHaveBeenCalledExactlyOnceWith(GitDiffReportModalComponent, { size: 'xl' });
+    });
+
+    it('should use reports from cache if available', fakeAsync(() => {
+        const diffForSubmissionsSpy = jest.spyOn(programmingExerciseService, 'getDiffReportForSubmissions');
+        const diffForSubmissionWithTemplateSpy = jest.spyOn(programmingExerciseService, 'getDiffReportForSubmissionWithTemplate');
+        component.previousSubmission = { id: 1 };
+        component.currentSubmission = { id: 2 };
+        component.exercise = { id: 3 } as ProgrammingExercise;
+        const cachedDiffReports = new Map<string, ProgrammingExerciseGitDiffReport>();
+        const expectedReport = {
+            id: 1,
+            leftCommitHash: 'abc',
+            rightCommitHash: 'def',
+            participationIdForLeftCommit: 1,
+            participationIdForRightCommit: 2,
+            programmingExercise: component.exercise,
+        } as ProgrammingExerciseGitDiffReport;
+        cachedDiffReports.set(JSON.stringify([1, 2]), expectedReport);
+        component.cachedDiffReports = cachedDiffReports;
+        component.ngOnInit();
+        component.exerciseIdSubject.next(1);
+        //tick 200 is needed because the observable uses debounceTime(200)
```

Comment:
I think there is some jest fake timer progression function 🤔 Maybe that could be useful here?

```suggestion
        // tick 200 is needed because the observable uses debounceTime(200)
```

Reply:
Using tick even with some values is the standard way we use in our entire Artemis test codebase, so I would just stick with it.